modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/macro/EmbedIframeMacroProcessor.java

```diff
@@ -24,37 +22,22 @@ public String getName()
     @Override
     public PortalResponse process( final MacroContext macroContext )
     {
-        final String body = HtmlHelper.unescape( macroContext.getBody() );
-        if ( isIframeHtml( body ) )
+        final String macroBody = macroContext.getBody();
+        final String body;
+        if ( isIframeHtml( macroBody ) )
         {
-            return generateResponse( body );
+            body = HtmlHelper.unescape( macroBody );
         }
-        return generateNonIframeResponse( macroContext );
-
-    }
-
-    private PortalResponse generateNonIframeResponse( final MacroContext macroContext )
-    {
-        if ( RenderMode.LIVE.equals( getRenderingMode( macroContext ) ) )
+        else
         {
-            return generateResponse( "" );
+            body = RenderMode.LIVE.equals( macroContext.getRequest().getMode() ) ? "" : "Expected an &lt;iframe&gt; element in Embed macro";
         }
-
-        return generateResponse( "Expected an &lt;iframe&gt; element in Embed macro" );
-    }
-
-    private PortalResponse generateResponse( final String body )
-    {
-        return PortalResponse.create().body( body ).build();
-    }
-
-    private boolean isIframeHtml( final String body )
-    {
-        return IFRAME_PATTERN.matcher( body ).matches();
+        return PortalResponse.create().contentType( MediaType.HTML_UTF_8 ).body( body ).build();
     }
 
-    private RenderMode getRenderingMode( final MacroContext macroContext )
+    private static boolean isIframeHtml( final String body )
     {
-        return macroContext.getRequest() == null ? RenderMode.LIVE : macroContext.getRequest().getMode();
+        // embed tag should be always escaped
+        return body.startsWith( "&lt;iframe" ) && body.endsWith( "&lt;/iframe&gt;" );
```

Comment:
body can be null. 
1. https://github.com/enonic/xp/blob/master/modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/postprocess/instruction/MacroInstruction.java#L190
2. https://github.com/enonic/xp/blob/master/modules/portal/portal-impl/src/main/java/com/enonic/xp/portal/impl/postprocess/instruction/Instruction.java#L32

Reply:
then `IFRAME_PATTERN.matcher( body ).matches();` would not work either

Reply:
👍🏻